### PR TITLE
 [SYCL][COMPAT] SYCLcompat dim and defs headers

### PR DESCRIPTION
**Pull request by Alcpz**
This is an implementation of SYCLcompat, the stand alone library proposed in https://github.com/intel/llvm/pull/9646.
Please, find the documentation there.

SYCLcompat has two primary goals:

- Improve the adoption of SYCL. This library is designed to provide a familiar programming interface that resembles other popular heterogeneous programming models. By reducing the learning curve, it enables developers to leverage SYCL's power and features more easily.
- Source-to-Source Translation Support. SYCLcompat is also designed to facilitate automatic source-to-source translation from other heterogeneous programming models to SYCL and offer a more standardized and consistent programming interface. This feature can significantly streamline the migration and integration of existing codebases into the SYCL ecosystem

The PR also includes tests. We were advised to include the tests as part of `sycl/unittests`, so we added a different mechanism to build them using `clang`.

As we stated in the docs PR, we are open to any suggestions, concerns, or improvements you may have, so please, let us know if you have any.

### CODE REVIEW

**Comment by Alcpz** [comment]
Documentation available in a different PR: https://github.com/intel/llvm/pull/9646

**Comment by aelovikov-intel** [comment]
Should we add a dedicated owner in `CODEOWNERS` file for this compat library?

File: sycl/unittests/syclcompat/Defs.cpp
Comment (aelovikov-intel) [inline sycl/unittests/syclcompat/Defs.cpp:27]: `DPCT`?

Comment (joeatodd) [reply sycl/unittests/syclcompat/Defs.cpp:27]: Well spotted, should be `Defs` I think.

File: sycl/unittests/CMakeLists.txt
Comment (Alcpz) [reply sycl/unittests/CMakeLists.txt:64]: We had a discussion and it was decided that it was best to have the headers, docs and tests in the relevant directories, rather than all together. For the tests, we were recommended to submit those within the existing `sycl/unittests` directory. But if you think it would be better to have them separated, we could also do that.

**Comment by Alcpz** [comment]
> Should we add a dedicated owner in `CODEOWNERS` file for this compat library?

I've added `sycl/doc/syclcompat/CodeOwners.rst` with the code owners for SYCLcompat. We will ask for the review group following this.

File: sycl/unittests/CMakeLists.txt
Comment (aelovikov-intel) [reply sycl/unittests/CMakeLists.txt:64]: + @steffenlarsen , @sergey-semenov

**Comment by Alcpz** [comment]
We have requested for the `syclcompat-lib-reviewers` group. The CODEOWNERS will be updated in a different request.

File: sycl/unittests/CMakeLists.txt
Comment (aelovikov-intel) [reply sycl/unittests/CMakeLists.txt:64]: Pinging @steffenlarsen and @sergey-semenov one more time :)

Comment (sergey-semenov) [reply sycl/unittests/CMakeLists.txt:64]: I think the current unittest location makes more sense considering the location of syclcompat headers and docs. Unless you're suggesting to move everything to a dedicated syclcompat directory, in which case, I don't really have a strong preference one way or the other.

**Comment by Alcpz** [comment]
@aelovikov-intel the entry in the [CODEOWNERS](https://github.com/intel/llvm/blob/5d928976f9e221a5a31ebdf805dc7bfd5e4a633b/.github/CODEOWNERS#L124) file has been merged.

**Review by joeatodd — APPROVED** [review]
LGTM

**Comment by joeatodd** [comment]
We're not really sure why @intel/dpcpp-doc-reviewers was requested for a review (possibly due to link with [this PR](https://github.com/intel/llvm/pull/9646)). Could someone have a quick look? Thanks!

**Comment by dm-vodopyanov** [comment]
The PR has all required approvals, waiting for CI completion to be able to merge.

**Comment by Alcpz** [comment]
@intel/llvm-gatekeepers CI is finally passing. We think it's ready to merge. Thanks!

**Pull request merged**
